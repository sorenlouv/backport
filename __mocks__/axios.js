const axios = jest.genMockFromModule('axios');
axios.get = jest.fn(() => {
  throw new Error('Forgot to mock axios.get');
});
axios.post = jest.fn(() => {
  throw new Error('Forgot to mock axios.post');
});
module.exports = axios;
