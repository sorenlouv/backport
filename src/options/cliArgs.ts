import yargs from 'yargs';
import { OptionsFromConfigFiles } from './config/config';

export type OptionsFromCliArgs = ReturnType<typeof getOptionsFromCliArgs>;
export function getOptionsFromCliArgs(
  configOptions: OptionsFromConfigFiles,
  argv: readonly string[]
) {
  const cliArgs = yargs(argv)
    .parserConfiguration({
      'strip-dashed': true,
      'strip-aliased': true,
      'boolean-negation': false,
    })
    .usage('$0 [args]')
    .wrap(Math.max(100, Math.min(120, yargs.terminalWidth())))
    .option('accessToken', {
      alias: 'accesstoken',
      description: 'Github access token',
      type: 'string',
    })
    .option('all', {
      default: configOptions.all,
      description: 'List all commits',
      alias: 'a',
      type: 'boolean',
    })
    .option('author', {
      default: configOptions.author,
      description: 'Show commits by specific author',
      type: 'string',
    })
    .option('maxNumber', {
      default: configOptions.maxNumber,
      description: 'Number of commits to choose from',
      alias: ['number', 'n'],
      type: 'number',
    })
    .option('dryRun', {
      default: false,
      description: 'Perform backport without pushing to Github',
      type: 'boolean',
    })
    .option('editor', {
      default: configOptions.editor,
      description: 'Editor to be opened during conflict resolution',
      type: 'string',
    })
    .option('fork', {
      default: configOptions.fork,
      description: 'Create backports in fork or origin repo',
      type: 'boolean',
    })
    .option('gitHostname', {
      hidden: true,
      default: configOptions.gitHostname,
      description: 'Hostname for Github',
      type: 'string',
    })
    .option('githubApiBaseUrlV3', {
      hidden: true,
      default: configOptions.githubApiBaseUrlV3,
      description: `Base url for Github's REST (v3) API`,
      type: 'string',
    })
    .option('githubApiBaseUrlV4', {
      hidden: true,
      default: configOptions.githubApiBaseUrlV4,
      description: `Base url for Github's GraphQL (v4) API`,
      type: 'string',
    })
    .option('mainline', {
      description:
        'Parent id of merge commit. Defaults to 1 when supplied without arguments',
      type: 'number',
      coerce: (mainline) => {
        // `--mainline` (default to 1 when no parent is given)
        if (mainline === undefined) {
          return 1;
        }

        // use specified mainline parent
        if (Number.isInteger(mainline)) {
          return mainline as number;
        }

        // Invalid value provided
        throw new Error(`--mainline must be an integer. Received: ${mainline}`);
      },
    })
    .option('multiple', {
      default: configOptions.multiple,
      description: 'Select multiple branches/commits',
      type: 'boolean',
    })
    .option('multipleCommits', {
      default: configOptions.multipleCommits,
      description: 'Backport multiple commits',
      type: 'boolean',
    })
    .option('multipleBranches', {
      default: configOptions.multipleBranches,
      description: 'Backport to multiple branches',
      type: 'boolean',
    })
    .option('noVerify', {
      default: configOptions.noVerify,
      description: 'Bypasses the pre-commit and commit-msg hooks',
      type: 'boolean',
    })
    .option('path', {
      default: configOptions.path,
      description: 'Only list commits touching files under the specified path',
      alias: 'p',
      type: 'string',
    })
    .option('prTitle', {
      default: configOptions.prTitle,
      description: 'Title of pull request',
      alias: 'title',
      type: 'string',
    })
    .option('prDescription', {
      default: configOptions.prDescription,
      description: 'Description to be added to pull request',
      alias: 'description',
      type: 'string',
    })
    .option('pullNumber', {
      conflicts: ['sha', 'sourcePRsFilter'],
      description: 'Pull request to backport',
      type: 'number',
      alias: 'pr',
    })
    .option('resetAuthor', {
      default: false,
      description: 'Set yourself as commit author',
      type: 'boolean',
    })
    .option('sha', {
      conflicts: ['pullNumber', 'sourcePRsFilter'],
      description: 'Commit sha to backport',
      type: 'string',
      alias: 'commit',
    })
    .option('sourcePRLabels', {
      default: configOptions.sourcePRLabels,
      description: 'Add labels to the source (original) PR',
      type: 'array',
      alias: 'sourcePRLabel',
    })
    .option('sourcePRsFilter', {
      conflicts: ['pullNumber', 'sha'],
      // default: configOptions.githubApiBaseUrlV4,
      description: `Filter source pull requests by a query`,
      alias: 'pr-filter',
      type: 'string',
    })
    .option('sourceBranch', {
      default: configOptions.sourceBranch,
      description: `List commits to backport from another branch than master`,
      type: 'string',
    })
    .option('verify', {
      description: `Opposite of no-verify`,
      type: 'boolean',
    })
    .option('targetBranches', {
      default: [] as string[],
      description: 'Branch(es) to backport to',
      type: 'array',
      alias: ['targetBranch', 'branch', 'b'],
      string: true, // ensure `6.0` is not coerced to `6`
    })
    .option('targetPRLabels', {
      default: configOptions.targetPRLabels,
      description: 'Add labels to the target (backport) PR',
      alias: ['labels', 'label', 'l'],
      type: 'array',
    })
    .option('upstream', {
      default: configOptions.upstream,
      description: 'Name of repository',
      alias: 'up',
      type: 'string',
    })
    .option('username', {
      default: configOptions.username,
      description: 'Github username',
      type: 'string',
    })
    .option('verbose', {
      default: false,
      description: 'Show additional debug information',
      type: 'boolean',
    })
    .alias('version', 'v')
    .alias('version', 'V')
    .help()
    .epilogue(
      'For bugs, feature requests or questions: https://github.com/sqren/backport/issues\nOr contact me directly: https://twitter.com/sorenlouv'
    ).argv;

  // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-unused-vars
  const { $0, _, verify, ...rest } = cliArgs;

  return {
    ...rest,
    accessToken: cliArgs.accessToken || configOptions.accessToken, // accessToken should not be displayed in yargs help menu
    branchLabelMapping: configOptions.branchLabelMapping, // not available as cli argument
    multipleBranches: cliArgs.multipleBranches || cliArgs.multiple,
    multipleCommits: cliArgs.multipleCommits || cliArgs.multiple,
    noVerify: verify ?? rest.noVerify, // `verify` is a cli-only flag to flip the default of `no-verify`
    targetBranchChoices: configOptions.targetBranchChoices, // not available as cli argument
  };
}
