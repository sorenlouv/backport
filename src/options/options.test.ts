import { OptionsFromCliArgs } from './cliArgs';
import {
  validateRequiredOptions,
  getOptions,
  BackportOptions,
} from './options';
import axios from 'axios';

describe('getOptions', () => {
  let axiosHeadSpy: jest.SpyInstance;
  let axiosPostSpy: jest.SpyInstance;
  let options: BackportOptions;

  beforeEach(async () => {
    jest.clearAllMocks();
    axiosHeadSpy = jest.spyOn(axios, 'head').mockReturnValueOnce(true as any);
    axiosPostSpy = jest.spyOn(axios, 'post').mockReturnValueOnce({
      data: {
        data: { repository: { defaultBranchRef: { name: 'myDefaultBranch' } } },
      },
    } as any);

    const argv = [
      '--branch',
      '6.0',
      '--branch',
      '6.1',
      '--upstream',
      'elastic/kibana',
    ];

    options = await getOptions(argv);
  });

  it('should check whether access token is valid', () => {
    expect(axiosHeadSpy).toHaveBeenCalledTimes(1);
    expect(axiosHeadSpy).toHaveBeenCalledWith(
      'https://api.github.com/repos/elastic/kibana',
      {
        auth: { password: 'myAccessToken', username: 'sqren' },
      }
    );
  });

  it('should get default repositry branch', () => {
    expect(axiosPostSpy).toHaveBeenCalledTimes(1);
  });

  it('should return options', async () => {
    expect(options).toEqual({
      accessToken: 'myAccessToken',
      all: false,
      githubApiBaseUrlV3: 'https://api.github.com',
      githubApiBaseUrlV4: 'https://api.github.com/graphql',
      author: 'sqren',
      backportCreatedLabels: [],
      branchChoices: [
        { checked: false, name: '6.0' },
        { checked: false, name: '5.9' },
      ],
      branches: ['6.0', '6.1'],
      fork: true,
      gitHostname: 'github.com',
      labels: [],
      multiple: false,
      multipleBranches: true,
      multipleCommits: false,
      prTitle: '[{baseBranch}] {commitMessages}',
      repoName: 'kibana',
      repoOwner: 'elastic',
      resetAuthor: false,
      sourceBranch: 'myDefaultBranch',
      username: 'sqren',
      verbose: false,
    });
  });
});

describe('validateRequiredOptions', () => {
  const validOptions: OptionsFromCliArgs = {
    accessToken: 'myAccessToken',
    all: false,
    githubApiBaseUrlV3: 'https://api.github.com',
    githubApiBaseUrlV4: 'https://api.github.com/graphql',
    author: undefined,
    backportCreatedLabels: [],
    branchChoices: [],
    branches: ['branchA'],
    commitsCount: 10,
    editor: 'code',
    fork: true,
    gitHostname: 'github.com',
    labels: [],
    multiple: false,
    multipleBranches: true,
    multipleCommits: false,
    path: undefined,
    prDescription: undefined,
    prTitle: 'myPrTitle',
    pullNumber: undefined,
    resetAuthor: false,
    sha: undefined,
    sourceBranch: 'mySourceBranch',
    upstream: 'elastic/kibana',
    username: 'sqren',
    verbose: false,
  };

  describe('should not throw', () => {
    it('when all options are valid and `branches` is given', () => {
      expect(() => validateRequiredOptions(validOptions)).not.toThrow();
    });

    it('when all options are valid and `branchChoices` is given', () => {
      expect(() =>
        validateRequiredOptions({
          ...validOptions,
          branchChoices: [{ name: 'branchA' }],
          branches: [],
        })
      ).not.toThrow();
    });
  });

  describe('should throw', () => {
    it('when accessToken is missing', () => {
      expect(() =>
        validateRequiredOptions({
          ...validOptions,
          accessToken: undefined,
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when accessToken is empty', () => {
      expect(() =>
        validateRequiredOptions({
          ...validOptions,
          accessToken: '',
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when both branches and branchChoices are missing', () => {
      expect(() =>
        validateRequiredOptions({
          ...validOptions,
          branchChoices: [],
          branches: [],
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when upstream is missing', () => {
      expect(() =>
        validateRequiredOptions({
          ...validOptions,
          upstream: undefined,
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when upstream is empty', () => {
      expect(() =>
        validateRequiredOptions({
          ...validOptions,
          upstream: '',
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when username is missing', () => {
      expect(() =>
        validateRequiredOptions({
          ...validOptions,
          username: undefined,
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when username is empty', () => {
      expect(() =>
        validateRequiredOptions({
          ...validOptions,
          username: '',
        })
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
