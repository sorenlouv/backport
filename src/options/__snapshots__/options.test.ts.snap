// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateRequiredOptions should throw when accessToken is empty 1`] = `
"Please update your config file: /myHomeDir/.backport/config.json.
It must contain a valid \\"username\\" and \\"accessToken\\".

Read more: https://github.com/sqren/backport/blob/434a28b431bb58c9a014d4489a95f561e6bb2769/docs/configuration.md#global-config-backportconfigjson"
`;

exports[`validateRequiredOptions should throw when accessToken is missing 1`] = `
"Please update your config file: /myHomeDir/.backport/config.json.
It must contain a valid \\"username\\" and \\"accessToken\\".

Read more: https://github.com/sqren/backport/blob/434a28b431bb58c9a014d4489a95f561e6bb2769/docs/configuration.md#global-config-backportconfigjson"
`;

exports[`validateRequiredOptions should throw when both branches and branchChoices are missing 1`] = `
"Invalid option \\"branches\\"

You can add it with either:
 - Config file: \\".backportrc.json\\". Read more: https://github.com/sqren/backport/blob/434a28b431bb58c9a014d4489a95f561e6bb2769/docs/configuration.md#project-config-backportrcjson
 - CLI: \\"--branches 6.1\\""
`;

exports[`validateRequiredOptions should throw when upstream is empty 1`] = `
"Invalid option \\"upstream\\"

You can add it with either:
 - Config file: \\".backportrc.json\\". Read more: https://github.com/sqren/backport/blob/434a28b431bb58c9a014d4489a95f561e6bb2769/docs/configuration.md#project-config-backportrcjson
 - CLI: \\"--upstream elastic/kibana\\""
`;

exports[`validateRequiredOptions should throw when upstream is missing 1`] = `
"Invalid option \\"upstream\\"

You can add it with either:
 - Config file: \\".backportrc.json\\". Read more: https://github.com/sqren/backport/blob/434a28b431bb58c9a014d4489a95f561e6bb2769/docs/configuration.md#project-config-backportrcjson
 - CLI: \\"--upstream elastic/kibana\\""
`;

exports[`validateRequiredOptions should throw when username is empty 1`] = `
"Please update your config file: /myHomeDir/.backport/config.json.
It must contain a valid \\"username\\" and \\"accessToken\\".

Read more: https://github.com/sqren/backport/blob/434a28b431bb58c9a014d4489a95f561e6bb2769/docs/configuration.md#global-config-backportconfigjson"
`;

exports[`validateRequiredOptions should throw when username is missing 1`] = `
"Please update your config file: /myHomeDir/.backport/config.json.
It must contain a valid \\"username\\" and \\"accessToken\\".

Read more: https://github.com/sqren/backport/blob/434a28b431bb58c9a014d4489a95f561e6bb2769/docs/configuration.md#global-config-backportconfigjson"
`;
